### PR TITLE
docs(aio): update ToH for CLI

### PR DESCRIPTION
**Pull request by Foxandxss**
Ready to go

### CODE REVIEW

File: aio/content/examples/toh-pt6/src/app/hero.service.ts [hunk 1]
```diff
@@ -1,87 +1,160 @@
 // #docplaster
-// #docregion , imports
-import { Injectable }    from '@angular/core';
-import { Headers, Http } from '@angular/http';
+// #docregion
+import { Injectable } from '@angular/core';
+// #docregion import-httpclient
+import { HttpClient, HttpHeaders }from '@angular/common/http';
```
Comment (Foxandxss) [inline aio/content/examples/toh-pt6/src/app/hero.service.ts]: missing space between }from

File: aio/content/tutorial/toh-pt5.md [hunk 2]
```diff
@@ -17,1356 +9,572 @@ There are new requirements for the Tour of Heroes app:
...
-
 <div class="l-sub-section">
 
-
-
-*Routing* is another name for *navigation*. The router is the mechanism for navigating from view to view.
-
-
+`--flat` puts the file in `src/app` instead of its own folder.<br>
+`--module=app` tells the CLI to register it in the `imports` array of the `AppModule.
```
Comment (Foxandxss) [inline aio/content/tutorial/toh-pt5.md]: missing ` at the end.

File: aio/content/tutorial/toh-pt5.md [hunk 3]
```diff
@@ -17,1356 +9,572 @@ There are new requirements for the Tour of Heroes app:
...
-You didn't add a `'Hero Detail'` link to the template because users
-don't click a navigation *link* to view a particular hero;
-they click a *hero name*, whether the name displays on the dashboard or in the heroes list.
+Here are the code files discussed on this page and your app should look like this <live-example></live-example>.
 
-You don't need to add the hero clicks until the `HeroDetailComponent`
-is revised and ready to be navigated to.
+{@a approutingmodule}
+{@a appmodule}
+#### _AppRoutingModule_ and _AppModule_
```
Comment (Foxandxss) [inline aio/content/tutorial/toh-pt5.md]: I completely understand the motive to create al this #### but they look horrible

**Comment by vicb** [comment]
missing approval

**Review by wardbell — APPROVED** [review]
LGTM and urgent to get aligned with v.5
We can do more polishing later

**Review by mhevery — APPROVED** [review]

**Comment by vicb** [comment]
@wardbell conflicting, could you please take a look

**Comment by Foxandxss** [comment]
rebased, waiting for travis

**Comment by Nicolas-Bouteille** [comment]
Hello, 
I believe that the code provided for the section Show the hero is not what it should be at that step.
The code is currently:
`<h2>{{ hero.name | uppercase }} Details</h2>
<div><span>id: </span>{{hero.id}}</div>
<div>
    <label>name:
      <input [(ngModel)]="hero.name" placeholder="name">
    </label>
</div>`
But at that step, hero is only a string. So such a code does not work yet and shows an error in the IDE: "Angular: Identifier 'name' is not defined. 'anonymous' does not contain such a member'
And the input does not bind with the Title.
Plus at that step the hero.id does not exist either so I guess the code that should be expected at that step is:
`<h2>{{ hero | uppercase }} Details</h2>
<div>
    <label>name:
      <input [(ngModel)]="hero" placeholder="name">
    </label>
</div>`

This gives more sense to the sentence at the end of the Create a Hero Class section:
"The page no longer displays properly because you changed the hero from a string to an object."

**Comment by Nicolas-Bouteille** [comment]
Just to be accurate, when creating a component with the CLI, the following line is now already added in app.module.ts:
`import { FormsModule } from '@angular/forms'; // <-- NgModel lives here`
the doc currently tells us to add it ourselves...

**Comment by Foxandxss** [comment]
Fixed the problem you mentioned. About FormsModule, it is not by default in a CLI app, you need to create it.

**Comment by Foxandxss** [comment]
@vicb PR is green but circleci fails and I cannot restart it.

**Comment by Nicolas-Bouteille** [comment]
In the section 6. Services >> Show Messages, the MessagesComponent's private CSS styles file is not mentioned. I noticed it in the final code review but I guess you would want to add a sentence that mentions it right?

**Comment by Foxandxss** [comment]
Maybe, but the tutorial doesn't focus much in the CSS. We can check it after we merge this. We need the update going, then we can polish it

**Comment by Nicolas-Bouteille** [comment]
Sure :)
So far I find the tutorial very well written and easy to follow!
Keeping routing for tomorrow though...

**Comment by Nicolas-Bouteille** [comment]
The code provided for 7. Routing >> Add routes, imports HeroDetailComponent instead of HeroesComponent.

**Comment by Foxandxss** [comment]
That is something to fix indeed. Thanks

**Comment by Nicolas-Bouteille** [comment]
7. Routing >> Add dashboard route : might be a good thing to remind the reader to import the DashboardComponent in app-routing.module.ts
`import { DashboardComponent } from './dashboard/dashboard.component';`
PHPStorm automatically adds it, but maybe some reader's IDE won't do it...

**Comment by Nicolas-Bouteille** [comment]
Same remark for 7. Routing >> Add a hero detail route
`import { HeroDetailComponent } from './hero-detail/hero-detail.component';`

**Comment by Nicolas-Bouteille** [comment]
7. Routing >> Add HeroService.getHero()
I noticed you used Template strings to display the message. 
Well, at first I did not notice and since I prefer writing the code myself instead of just pasting it so that I remember it better, I used regular quotes. So the variable id would not be correctly replaced in the message. Then I noticed the special quotes that I did not know about and did a bit of searching. Only then did I come across the explanation of Template strings...
So I guess it would be nice to add something like:
"In case you did not notice, we used backticks \` called Template strings in order to print the id variable inside our message. Read more about Template strings here: https://basarat.gitbooks.io/typescript/docs/template-strings.html"

**Comment by Foxandxss** [comment]
Isn't there a explanation of the backstick somewhere? But I can see that happening.

**Comment by Nicolas-Bouteille** [comment]
Maybe later, but at that step nope... So if the reader, like me, does not know about it, he/she might not understand or notice.

**Comment by Nicolas-Bouteille** [comment]
I find it weird that the Simulate the Web API is an Appendix.
When you give links to files at the end of the section, it's a bit like for CSS, we don't really know if we should do it right away or not... Me, I assume it's not too important or you would include it right away... so it can wait at the end...
But at the end of the next section Get Heroes with HttpClient, you guys say:
"Refresh the browser. The hero data should successfully load from the mock server."
So this means you expect us to do the Appendix right away... so why put it as an Appendix and not a section? I'm confused.

**Comment by Foxandxss** [comment]
Let's bring @wardbell to the conversation, he created that new sections.

**Comment by Nicolas-Bouteille** [comment]
Also, `import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';`
throws an error. I guess we need to download the module first from https://github.com/angular/in-memory-web-api and put it somewhere... 
But so far we have not learnt how to do it (download an external module), so I am personally stuck here. The zip file contains lots of files. Should we put them all in a custom folder angular-in-memory-web-api ? Should we npm install it? 
Maybe there should be some guidance here about it?

**Comment by Foxandxss** [comment]
you need to `npm install -S angular-in-memory-web-api` so you will be able to add that line.

We will update the doc

**Comment by Nicolas-Bouteille** [comment]
Ok I see this adds a dependency in package.json file. Suggestion: maybe this could be one of devDependencies instead?

**Comment by Nicolas-Bouteille** [comment]
Also, just a Warning that might be useful for @wardbell to know:
```
npm WARN codelyzer@3.2.2 requires a peer of @angular/compiler@^2.3.1 || >=4.0.0-beta <5.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN codelyzer@3.2.2 requires a peer of @angular/core@^2.3.1 || >=4.0.0-beta <5.0.0 but none is installed. You must install peer dependencies yourself.
```

**Comment by Foxandxss** [comment]
Since the app will go to production with that, it is a dependency. But well, for the tutorial is not that important.

**Comment by Nicolas-Bouteille** [comment]
Ok I see!

**Comment by Nicolas-Bouteille** [comment]
8. HTTP >> Add HeroService.addHero()
TSLint reports something I don't understand:
shadowed name
Also, what's with the 'w/' ? Does it mean something or is it a typo?

**Comment by Foxandxss** [comment]
`w/` is an english shorthand which means "with", there is also `w/o` which is `without`. We should remove that I guess.

**Comment by Nicolas-Bouteille** [comment]
ok thx

**Comment by Nicolas-Bouteille** [comment]
8. HTTP >> Search by name.
I was wondering why the variable `heroes$: Observable<Hero[]>;` ends with a dollar sign?

**Comment by Foxandxss** [comment]
That is explained in the text.

**Comment by Nicolas-Bouteille** [comment]
Oups! Missed that! Sorry

**Comment by Nicolas-Bouteille** [comment]
Alright I'm done with the Tutorial! Yeah!

One final note that I had regarding the last section (HTTP), compared to the previous ones, it looked more like Test Driven Development. I mean that, in the previous sections, we built components and then we included them somewhere. In the last section, we include stuff that don't exist yet, so the IDE throws error, and finally when everything is done, no more errors. Just like TDD. 
Personally I like TDD, but I prefer creating components first, so that the IDE autocomplete works and does not throw errors... So in the last part I have to confess I chose to create the components and methods first before including or using them. I'm not saying that the last part should change, I'm just saying that it is different from the previous sections and that I personally felt more comfortable creating stuff before mentioning them somewhere else in my code... So just a personal feedback...

Anyway, thanks a lot for this great Tutorial guys!

**Comment by wardbell** [comment]
@Nicolas-Bouteille 

Thank you for your most generous and thorough review.  I've addressed your concerns as best I can in the following ways.

1. The missing instructions to add CSS in "Hero Service".  Definitely needed although not where you asked for them. There are no messages _at that particular point_.  I've added instructions down where you modify the `MessagesComponent` to show messages.

1. Fixed the incorrect `import` instruction in "Routing > Add routes"

1. Added the two missing "Routing" `import`instructions that you mentioned.

1. Added the backtick comment in "Routing".

1. I struggled with where to put the "In memory web API".  I have come around to your position that it belongs in the tutorial flow. I've rewritten accordingly while striving to make clear that it has nothing to do with `HttpClient` and you can skip this step if you're just reading the page.

1. Add instruction for installing the `angular-in-memory-web-api` npm package. I'm not sure why you get those `codelyzer` warnings. I don't think this package is the cause.

I would like to do something about the fact that the app keeps failing while you're building it.
It's a process of breaking and fixing.

You suggested that creating the components first would alleviate the problem. I don't think so. Adding methods to the components that refer to services that don't exist yet leaves the app broken. Jumping between component and service and back again might stop some of the breakage but not all of it and I think it would be more confusing ... certainly more confusing for someone just reading the tutorial.

Therefore, having no good answer, I'm leaving it as is for now.

Thanks again!

**Comment by Nicolas-Bouteille** [comment]
I am quite amazed to see the reactivity of the Angular community! 
Congrats...

**Comment by aaaaarrrgghhh** [comment]
I'm having the same error that Nicolas mentioned on nov 4 above, but I don't see a resolution of that.
 its at this link in the tutorial https://angular.io/tutorial/toh-pt6#add-heroserviceaddhero

The addHero method results in TSLint complaining of a "Shadowed name: hero" for the declaration in "tap". I changed the variable name and its all fine now.
`tap((hero1: Hero) => this.log(`added hero w/ id=${hero.id}`)),`

**Comment by jenniferfell** [comment]
@mhevery Would you do me a favor and take a look at the comment above from a few days ago. The recommended change of variable name has not been made to the current released TOH. 

Q1: Can you reproduce/verify the problem described in that comment? Or should I ask someone else (@kapunahelewong?)?
Q2: If the problem still exists, is the right procedure to reopen this PR or to create a new doc issue with the comment and proposed resolution?

Thanks.

**Comment by Foxandxss** [comment]
I can't reproduce the problem

**Comment by kapunahelewong** [comment]
This error that @aaaaarrrgghhh mentioned sounds like it's a variable that is inadvertently declared twice. Might that be the case, @aaaaarrrgghhh?

@jenniferfell This would be a good candidate for a new issue if @aaaaarrrgghhh can't find the cause for the error. I can't replicate the issue on my end. Everything works as expected for me.

**Comment by aaaaarrrgghhh** [comment]
Yes, I think the variable name 'hero' is defined twice in the method
` /** POST: add a new hero to the server */
addHero (**hero: Hero**): Observable<Hero> {
  return this.http.post<Hero>(this.heroesUrl, hero, httpOptions).pipe(
    tap((**hero1: Hero**) => this.log(`added hero w/ id=${hero.id}`)),
    catchError(this.handleError<Hero>('addHero'))
  );
}`

**Pull request merged**